consensus/src/dag/dag_store.rs

```diff
@@ -0,0 +1,268 @@
+        #[derive(Serialize)]
+        struct NodeWithoutDigest<'a> {
+            epoch: u64,
+            round: Round,
+            author: Author,
+            timestamp: u64,
+            payload: &'a Payload,
+            parents: &'a Vec<NodeMetadata>,
+        }
+
+        impl<'a> CryptoHash for NodeWithoutDigest<'a> {
+            type Hasher = NodeHasher;
+
+            fn hash(&self) -> HashValue {
+                let mut state = Self::Hasher::new();
+                let bytes = bcs::to_bytes(&self).expect("Unable to serialize node");
+                state.update(&bytes);
+                state.finish()
+            }
+        }
+
+        let node_with_out_digest = NodeWithoutDigest {
+            epoch,
+            round,
+            author,
+            timestamp,
+            payload,
+            parents,
+        };
+        node_with_out_digest.hash()
+    }
+
+    pub fn digest(&self) -> HashValue {
+        self.metadata.digest
+    }
+
+    pub fn metadata(&self) -> NodeMetadata {
+        self.metadata.clone()
+    }
+}
+
+/// Quorum signatures over the node digest
+#[derive(Clone)]
+pub struct NodeCertificate {
+    digest: HashValue,
+    signatures: AggregateSignature,
+}
+
+impl NodeCertificate {
+    pub fn new(digest: HashValue, signatures: AggregateSignature) -> Self {
+        Self { digest, signatures }
+    }
+}
+
+#[derive(Clone)]
+pub struct CertifiedNode {
+    node: Node,
+    certificate: NodeCertificate,
+}
+
+impl CertifiedNode {
+    pub fn new(node: Node, certificate: NodeCertificate) -> Self {
+        Self { node, certificate }
+    }
+}
+
+impl Deref for CertifiedNode {
+    type Target = Node;
+
+    fn deref(&self) -> &Self::Target {
+        &self.node
+    }
+}
+
+pub struct Dag {
+    nodes_by_digest: HashMap<HashValue, Arc<CertifiedNode>>,
+    nodes_by_round: BTreeMap<Round, Vec<Option<Arc<CertifiedNode>>>>,
+    author_to_index: HashMap<Author, usize>,
+    highest_unlinked_nodes_by_author: Vec<Option<Arc<CertifiedNode>>>,
+}
+
+impl Dag {
+    pub fn new(author_to_index: HashMap<Author, usize>, initial_round: Round) -> Self {
+        let mut nodes_by_round = BTreeMap::new();
+        let num_nodes = author_to_index.len();
+        nodes_by_round.insert(initial_round, vec![None; num_nodes]);
+        Self {
+            nodes_by_digest: HashMap::new(),
+            nodes_by_round,
+            author_to_index,
+            highest_unlinked_nodes_by_author: vec![None; num_nodes],
+        }
+    }
+
+    fn lowest_round(&self) -> Round {
+        *self
+            .nodes_by_round
+            .iter()
+            .next()
+            .map(|(round, _)| round)
+            .unwrap_or(&0)
+    }
+
+    fn highest_round(&self) -> Round {
+        *self
+            .nodes_by_round
+            .iter()
+            .next_back()
+            .map(|(round, _)| round)
+            .unwrap_or(&0)
+    }
+
+    pub fn add_node(&mut self, node: CertifiedNode) -> anyhow::Result<()> {
```

Comment:
Where do we verify the signature? Also, it seems like the `signature` stored in certified node is never being used?

Reply:
commented above too, self-contained checks will be verified outside of this, here we validate context-aware properties

Reply:
why do you need the `signature` to be stored if this is verified outside?

Reply:
it needs to reply the data to other nodes that request for missing nodes (and the remote needs to be able to verify the signatures)